### ISSUE TITLE
Logical organisation of OCFL storage root

### ISSUE THREAD

**Issue by tomwrobel** [OCFL/spec#249]
I appreciate that you've almost certainly gone over this many times, but I think it's still important to have some formal way of describing the logical organisation of the OCFL storage root.

I realise in spec/4.4 Extensions it says that you can provide the structure of the storage hierarchy, and in implementation/rebuildability there is a suggestion that this should be done. However, I don't think this goes far enough.

It might be made optional, but I think there should be an OCFL approved way to describe the storage hierarchy in relation to the object ID. 

Reasons:

1. It's good practice as this promotes long term sustainability. The storage hierarchy adopted may be obvious now, but that doesn't mean it will be so to future users of the data. An unexplained hierarchy is an additional problem future users will have to puzzle through.
2. Not doing so is a missed opportunity, as there is a chance to guide users towards a standard way to do this at the very start of the project.
3. It has practical benefits: In most implementations, I expect there would be a logical way to derive an object's location within the OCFL file system from its ID (otherwise an application would have to maintain a mapping externally). 
  3.1. Having a way to provide a formal definition of that relationship that is properly externally documented is very useful. 
  3.2 Although not necessarily in the first pass, an OCFL-approved statement (via NAMASTE?) could be used by a validation system to ensure that the Object ID/Object path relationship was intact.

Again, sorry if I'm being a pain, but this seemed to me to be something I actively want to be able to define in a compatible and sustainable way.

**Comment by ahankinson** [comment]
Thanks Tom! You’re right, we have discussed this extensively; see #214 and #138. The gist of it is that it would certainly be a valuable thing to have, but that it would be better expressed in a complimentary spec, either on its own or as part of the pairtree spec.

The issues are thus: fundamentally that there is no satisfactory way to express content disposition succinctly with a namaste file. Embedding a regex in the filename to determine a pattern with which to validate and parse the paths will almost certainly run into problems with file names. Creating a naming scheme (e.g. 1=2-2-3-n) would be more preferable, but developing a naming scheme that covers all cases would be difficult. Some orgs will undoubtedly need to strip special characters (slashes, colons) from their paths, while others will be able to use them literally. Trying to encode all of these possibilities is beyond the scope of the OCFL effort, but the intent of the extensions mechanism is that if such a spec does emerge, OCFL will not get in the way.

Pairtree will also not be used in every implementation of OCFL. Object stores do not have the same limitations as traditional file systems, and storing them in a flat list is perfectly acceptable.

Finally, we do have a requirement that the end nodes of every file path MUST be an OCFL object; that is, intermediate nodes in a storage root Must not contain anything other than child directories. This essentially eliminates the danger of hidden content, as it is always possible for a human to navigate the hierarchy and determine the layout. For software clients then it is trivial to compute and store the disposition, or for it to be a configuration parameter for clients. This gives institutions more flexibility for arranging their storage roots according to local considerations (identifier format being the biggest). For OCFL it doesn’t matter, as long as it is *consistent*.

**Comment by tomwrobel** [comment]
I understand all the reasons why it can't be a mandated setting. I also understand why all cases can't be covered. I just think there's a middle ground here! This is, for me, a big hole in documenting the structure of an OCFL filesystem. 

> it is trivial to compute and store the disposition the first time it reads from the store

Only once you've read the object. Which you have to find first. A full directory traversal is a very expensive task.

> or for it to be a configuration parameter for clients

I like the idea of an OCFL filesystem as being self contained. I wouldn't want to store key configuration parameters outside of the storage root. If it's an important enough parameter, we should be able to store it in the filesystem.

Reading #214 et al, I've seen the idea of a pointer to a special object. For me, this invites problems. I'd rather not have some objects in a filesystem as meta-objects. The clutter gets very annoying (Fedora3 repositories are often very guilty of this, and it costs us a lot of execution time filtering stuff out).

Would a URI for an existing standard be acceptable here? OCFL depends on external standards elsewhere. You could fit that in a Namaste file. Otherwise, a *recommended* directory-structure file (with contents) would be sufficient. Honestly, I don't think directory structure descriptions should be optional at all (allowing for directory-structure=none), but that's just me.

**Comment by zimeon** [comment]
I agree with @ahankinson's comment that it is not possible to succinctly describe possible arrangements in some machine readable way. I find the idea of providing a link to an external spec as @tomwrobel suggests appealing but that is not specific enough (e.g. if one linked to the Pairtree spec., would that mean an encapsulated or unencapsulated example; with pairs, triples or quads?). My favored solution at present is to have some identifier for a particular structure. This could either be a URI  (in which case one might hope it resolves to a description, but URIs are ugly in NAMASTE so maybe consider other options like a JSON file), or else a simple string (works with NAMASTE `1=something`, would then have to point to a registry to avoid collisions).

**Comment by zimeon** [comment]
Proposal from https://github.com/OCFL/spec/wiki/2018.11.07-Editors-Meeting to insert into https://ocfl.io/draft/spec/#root-structure something along the lines of:

> An OCFL Storage Root MAY contain a JSON file named `ocfl_map.json` to describe the arrangement of directories and OCFL objects under the storage root. If present, this JSON document MUST include the following two keys in the root JSON object:
>   * `uri` - A URI identifying the precise arrangement of directories and OCFL objects under the storage root. If a URL is used then it SHOULD resolve to a detailed specification of the arrangement.
>   * `description` - A human readable description of the arrangement of directories and OCFL objects under the storage root.

and the structure would also be updated:

> The following example OCFL Storage Root represents the minimal set of files and folders:
>
> [storage_root]
>    ├── 0=ocfl_1.0
>    ├── ocfl_1.0.txt (optional)
>   └── ocfl_map.json (optional)

**Comment by ahankinson** [comment]
Bikeshedding: Should it be `id` instead of `uri` to be consistent with `id` in inventories?

**Comment by ahankinson** [comment]
Or is that a foolish consistency?

**Comment by ahankinson** [comment]
On further reflection, `uri` is better.
